### PR TITLE
Refactor train.py and model.py

### PR DESCRIPTION
**Pull request by timcanby**

### CODE REVIEW

File: MoleculeVAE-reImplementation/Lightning_version/logs/my_exp_name/version_0/metrics.csv [hunk 1]
```diff
@@ -0,0 +1,18 @@
+train_loss,epoch,step
```
Comment (ebakker1) [inline MoleculeVAE-reImplementation/Lightning_version/logs/my_exp_name/version_0/metrics.csv:1]: I don't think this file should be part of the repo.

File: MoleculeVAE-reImplementation/train.py [hunk 2]
```diff
@@ -1,96 +1,330 @@
...
+    '''
+            - Inputs: params (Use hyperparameters.load_params() to load params from expParam.json)
+        - Returns:
+        A pandas DataFrame include follow items:
+            -train_set_final,
+            -valid_set_final,
+            -test_dataset,
+        '''
+    if params['do_prop_pred']:
+        X_train, X_test, Y_train, Y_test, character_index_lookup_dict = load_dataset(params)
```
Comment (ebakker1) [inline MoleculeVAE-reImplementation/train.py]: I'm kind of confused here. `load_dataset` has multiple outputs, but when loading from file the return of `load_dataset` seems to be a dataframe, so would it then break?

File: MoleculeVAE-reImplementation/train.py [hunk 3]
```diff
@@ -1,96 +1,330 @@
...
-    def __getitem__(self, idx):
-        return self.X[idx], self.y[idx]
 
 def vae_loss(x_decoded_mean, x, z_mean, z_logvar):
     xent_loss = nn.MSELoss()
-    vecloss=xent_loss (x_decoded_mean,x)
+    vecloss = xent_loss(x_decoded_mean.to(dtype=torch.float32, device=device), x.to(dtype=torch.float32, device=device))
     kl_loss = -0.5 * torch.sum(1 + z_logvar - z_mean.pow(2) - z_logvar.exp())
-    return vecloss + kl_loss
+    record_list.append()
```
Comment (ebakker1) [inline MoleculeVAE-reImplementation/train.py]: This is a bit odd - it seems like nothing is getting appended which I would expect to throw and error. Maybe this is a sign you're not using this function anymore and you would be better to delete it.

File: MoleculeVAE-reImplementation/train.py [hunk 4]
```diff
@@ -1,96 +1,330 @@
...
+            pre_loss_ca = nn.MSELoss()
+            pre_loss = pre_loss_ca(pre, label)
+            loss = calculate_loss_by_type(output, data, mean, logvar, pre_loss, epochs, 'train')
             loss.backward()
             train_loss += loss
             optimizer.step()
-            if batch_idx % 1000 == 0:
-                torch.save(model.state_dict(), 'Weights/param.pth')
-                print(f'{epochs} / {batch_idx}\t{loss:.4f}')
+            if batch_idx % len(train_set_final) == 0:
```
Comment (ebakker1) [inline MoleculeVAE-reImplementation/train.py]: Putting inside this `if` is a bit odd, as it just means it will get triggered at the end of the loop. Why not put it in the epoch look but outside the train set loop?

File: MoleculeVAE-reImplementation/train.py [hunk 5]
```diff
@@ -99,17 +333,40 @@ def train(epochs):
...
-        df = pd.DataFrame(list, columns=['epoch','logP', 'qed', 'SAS', 'mae_logP', 'mae_qed', 'mae_SAS'])
-        df.to_csv('trainingprocess.csv')
+        training_data_dic = load_data_by_task(params)
+        train_set_final = training_data_dic['train_set_final'].to_numpy()
+        train_loss, valid_loss = train(epoch, training_data_dic['train_set_final'].tolist(),
+                                       training_data_dic['valid_set_final'].tolist())
+        test_loss = test(training_data_dic['test_dataset'].tolist(), epoch)
+        # print([epoch, train_loss, valid_loss, test_loss])
 
+    '''separate_prediction_list For evaluate the prediction results on test set of each epoch'''
```
Comment (ebakker1) [inline MoleculeVAE-reImplementation/train.py]: These strings seem to actually be comments, since they're not used in the code. If that's so it's probably best to write them as:
```
# separate_prediction_list For evaluate the prediction results on test set of each epoch
```

File: MoleculeVAE-reImplementation/train.py [hunk 6]
```diff
@@ -1,96 +1,330 @@
...
+dt_string = datetime.utcnow().strftime("%d/%m/%Y %H:%M:%S")
+print("date and time =", dt_string)
+
+'''Lists for recording experimental results'''
+# loss_record_list for recording the loss terms [epoch,vae_loss, reconstruction_loss, prediction_loss,process_type]
+record_list = []
+# performance on separate prediction tasks [epoch,'logP_pre', 'qed_pre', 'SAS_pre', 'mae_logP', 'mae_qed', 'mae_SAS']
+separate_prediction_list = []
+# list for record annealing performance [epoch,test_total_loss_with_annealing,test_total_loss_without_annealing,annealing_weights,process_type]
+list_for_record_annealing_performance = []
```
Comment (ebakker1) [inline MoleculeVAE-reImplementation/train.py]: Here you are using these global (or at least Module level) variables to record losses over time. I understand the temptation but I think it's better to have them as normal (local variables) and pass them out of your `train` and `test` functions.

**Review by ebakker1 — CHANGES_REQUESTED** [review]
Codes already looking much better Kangying! I've left some comments but I don't think we need to implement them at this time - They're just for information and discussion. I think we're best just to remove the files that shouldn't be part of the repo (`__pycache__`, `.pyc` and maybe the logging folder) and merge. We can come back for another refactor in the future if we want to.


* There's a few `.pyc` files in the repo. These shouldn't be part of the git repo and should be removed. You can stop them getting added by including a `.gitignore` file like [this one](https://github.com/recursiveai/tenpureto/blob/main/templates/app/.gitignore) from our internal template or [this one](https://github.com/github/gitignore/blob/main/Python.gitignore) from a public repo. Similarly for the `__pycache__` folder. 
* There seems to be a lot of overlap between the lightning version of `dataloader`. Is that right? Is there subtle differences I can't see?

File: MoleculeVAE-reImplementation/Lightning_version/logs/my_exp_name/version_0/metrics.csv (hunk 1 above)
Comment (timcanby) [reply MoleculeVAE-reImplementation/Lightning_version/logs/my_exp_name/version_0/metrics.csv:1]: this maybe the default log file of Lightning when I try to use  CSVLogger("logs", name='')

File: MoleculeVAE-reImplementation/train.py (hunk 3 above)
Comment (timcanby) [reply MoleculeVAE-reImplementation/train.py]: sorry it was a mistake I forgot to delete it after some modifications

File: MoleculeVAE-reImplementation/train.py (hunk 2 above)
Comment (timcanby) [reply MoleculeVAE-reImplementation/train.py]: Sorry ,`load_dataset ` here maybe a function defined in ` dataloader.py ` file so it should become multiple outputs.

Comment (timcanby) [reply MoleculeVAE-reImplementation/train.py]: Oh I apologize that I found the dataloader.py here is not the updated version!!  I will update them!!sorry!!

File: MoleculeVAE-reImplementation/train.py (hunk 6 above)
Comment (timcanby) [reply MoleculeVAE-reImplementation/train.py]: Thank you for telling me this,I will try some other ways to record them

Comment (timcanby) [reply MoleculeVAE-reImplementation/train.py]: I've tried to summarize them to a class and record the loss value, is that a solution could be used?
